Partial Rewrite

gaymeowing opened the pull request:
* Switch from Vector3 methods to vector library methods
* Improve types
* Export types
* Allow for other WorldRoots besides Workspace
* Optimizations on getGeometry (didn't benchmark but should be slightly better atleast, given its now using vector library methods that are fast-called)
* Vertex, Edge, and Face are now "classes" as in they have a frozen metatable that stores the type field
* Add .luaurc
* Use .luau instead of .lua given github has added syntax highlighting for luau
* Freeze the module table

gaymeowing committed:
Partial Rewrite

gaymeowing committed:
forgot to pass around worldRoot

gaymeowing committed:
switch assert in getShape to use error

gaymeowing committed:
freeze module

gaymeowing committed:
oops

stravant commented:
Why did you parameterize the exported types by ID type?

That seems like it would create considerable frustration for consumers of the library as they would have to add an unnecessary type parameter every time they want to use the exported types.

gaymeowing commented:
> That seems like it would create considerable frustration for consumers of the library as they would have to add an unnecessary type parameter every time they want to use the exported types.

They don't because the ID generic is by default nil (just doing `Edge` is ok) unless its set by the user ie `Edge<number>` , but its there because of `getGeometry` using those types with the id field as `number` and blackbox also uses edge but it doesn't set the id field. 
For the others besides edge, its just future-proofing in case some other method in future also has ids like getGeometry does.

gaymeowing committed:
* reformat string enum types
* fix return type for findNearestInterestingPoint
* make findNearestInterestingPoint & exploreEdge take in a raycastParams arg rather than making a new raycastParams for both
* remove unused variables from blackboxFindClosestMeshEdge
* remove intersectRayPlane as its not used as a function anywhere besides intersectRayPlanePoint, its being kept as a comment now

gaymeowing committed:
added doc comments to blackboxFindClosestMeshEdge and getGeometry

gaymeowing committed:
add .gitignore for rojo sourcemaps

gaymeowing committed:
Clean up getGeometry
* remove second argument as its entirely unsued
* remove the if case for spheres & cylinders as thats handled by the first case

gaymeowing committed:
Correct README to show the new types & also state that the cframe is the second arg of getGeometry not the third now

gaymeowing commented:
Did some more improvements I missed, and some easy optimizations like reusing a RaycastParams object.
Also removed the second case of the shape if statement in `getGeometry` as its unused, and removed the `hit` arg because its only used in the second case.

The doc comments for the functions use the same description as the README, although I didn't include the performance note for `blackboxFindClosestMeshEdge` because raycasts are now very cheap.

I'll be going over gapfill & the other geomtools plugins to make them use the new version of roblox-geometry as this has became a breaking change w the second arg of `getGeometry` being removed. 
I also may do some more changes to this module if some of the fields for edge & face become unneeded, as it'd be an easy way to reduce memory usage of this module by a bit. But that'd be a diffrent pr, as this pr is more so focused on the bulk of stuff & having the types be exported so gapfill can have its codebase be fully typed.
